### PR TITLE
Recursive URL loader

### PR DESCRIPTION
**Pull request by rlancemartin**
We may want to process load all URLs under a root directory.

For example, let's look at the [LangChain JS documentation](https://js.langchain.com/docs/).

This has many interesting child pages that we may want to read in bulk.

Of course, the `WebBaseLoader` can load a list of pages. 

But, the challenge is traversing the tree of child pages and actually assembling that list!
 
We do this using the `RecusiveUrlLoader`.

This also gives us the flexibility to exclude some children (e.g., the `api` directory with > 800 child pages).

### CODE REVIEW

**Comment by kdcokenny** [comment]
https://github.com/hwchase17/langchain/commit/301318e803991658c95bb6f8f5d94acec39e0ef9

I reformatted your code to pass linting checks. Idk how to add the changes here, so feel free to just copy and paste the changes lol

**Comment by rlancemartin** [comment]
> 301318e

Thanks! Appreciate it. Just pulled in those changes and tested (all worked). Saved me some work :) 

BTW, did you test it on some use cases of your own (if you have any)?

Curious if any bugs on various types of websites; I have not yet tested broadly.

File: langchain/document_loaders/recursive_url_loader.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+from bs4 import BeautifulSoup
+
+from langchain.docstore.document import Document
+from langchain.document_loaders.base import BaseLoader
+
+
+class RecusiveUrlLoader(BaseLoader):
+    """Loader that loads all child links from a given url."""
+
+    def __init__(self, url: str, exclude_dirs: Optional[str] = None) -> None:
```
Comment (eyurtsev) [inline langchain/document_loaders/recursive_url_loader.py:14]: Maybe exclude paths or patterns for the variable mame? Dirs remind me of the file system

**Review by eyurtsev — APPROVED** [review]
Looks pretty reasonable to me, tests would be nice if we can use mock requests or some other way to mock

File: langchain/document_loaders/recursive_url_loader.py (hunk 1 above)
Comment (rlancemartin) [reply langchain/document_loaders/recursive_url_loader.py:14]: Ya, was thinking about this too. Will look into it!

**Comment by yadavsahil197** [comment]
Hi @rlancemartin this commit just caused an import error.
```
packages/langchain/document_loaders/recursive_url_loader.py:5: in <module>
    from bs4 import BeautifulSoup
E   ModuleNotFoundError: No module named 'bs4'
```

**Comment by rlancemartin** [comment]
> Hi @rlancemartin this commit just caused an import error.
> 
> ```
> 
> packages/langchain/document_loaders/recursive_url_loader.py:5: in <module>
> 
>     from bs4 import BeautifulSoup
> 
> E   ModuleNotFoundError: No module named 'bs4'
> 
> ```

Yes, it should be a local import.

Will push a fix shortly.

**Comment by F2EVarMan** [comment]
image
@rlancemartin

**Comment by rlancemartin** [comment]
> image
> 
> @rlancemartin 

Hmm. Strange. The PR added lazy_load this to the web loader.

**Pull request merged**